gazelle/parser/tokenizer.go

```diff
@@ -0,0 +1,495 @@
+// Package parser provides utilities useful for parsing a sass file.
+package parser
+
+// I used
+// https://blog.gopheracademy.com/advent-2014/parsers-lexers/
+// heavily for reference when creating this file.
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+	"unicode/utf8"
+)
+
+var eof = rune(0)
+
+type Token interface {
+	String() string
+	Type() string
+}
+
+func isHexDigit(ch rune) bool {
+	// A digit, or a code point between U+0041 LATIN CAPITAL LETTER A (A) and U+0046 LATIN CAPITAL LETTER F (F) inclusive, or a code point between U+0061 LATIN SMALL LETTER A (a) and U+0066 LATIN SMALL LETTER F (f) inclusive.
+	return isDigit(ch) || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F')
+}
+func isDigit(ch rune) bool {
+	// A code point between U+0030 DIGIT ZERO (0) and U+0039 DIGIT NINE (9) inclusive.
+	return (ch >= '0' && ch <= '9')
+}
+func isLetter(ch rune) bool {
+	// An uppercase letter or a lowercase letter.
+	return isLowercaseLetter(ch) || isUppercaseLetter(ch)
+}
+func isLowercaseLetter(ch rune) bool {
+	// A code point between U+0061 LATIN SMALL LETTER A (a) and U+007A LATIN SMALL LETTER Z (z) inclusive.
+	return ch >= 'a' && ch <= 'z'
+}
+func isUppercaseLetter(ch rune) bool {
+	// A code point between U+0041 LATIN CAPITAL LETTER A (A) and U+005A LATIN CAPITAL LETTER Z (Z) inclusive.
+	return ch >= 'A' && ch <= 'Z'
+}
+func isNonASCIICodePoint(ch rune) bool {
+	// A code point with a value equal to or greater than U+0080 <control>.
+	return ch >= '\u0080'
+}
+func isNameStartCodePoint(ch rune) bool {
+	// A letter, a non-ASCII code point, or U+005F LOW LINE (_).
+	return isLetter(ch) || isNonASCIICodePoint(ch) || ch == '_'
+}
+func isNameCodePoint(ch rune) bool {
+	// A name-start code point, a digit, or U+002D HYPHEN-MINUS (-).
+	return isNameStartCodePoint(ch) || isDigit(ch) || ch == '-'
+}
+
+// New creates a scanner that reads the provided io.Reader.
+func New(r io.Reader) *Scanner {
+	return &Scanner{r: bufio.NewReader(r)}
+}
+
+// Scanner is responsible for reading an input file.
+type Scanner struct {
+	r *bufio.Reader
```

Comment:
I did it this way so I wouldn't have the whole file in memory. I definitely considered doing it (so many times when I was writing that peek code), but was thinking back to some of the more unwieldy CSS files I've seen generated and their enormous length.

Out of curiosity, do you have data to support the claim that reading the whole file is faster? This feels like a real gotcha in golang that the buffered reader on a file is less performant than reading the whole file into a buffer yourself. I guess it might be `mmap`ed and read that way which I guess would be faster than sequential reads, but I can't quite make the jump in my head.